src/asus.c

```diff
@@ -0,0 +1,429 @@
+	ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_SPECIAL);
+	ratbag_button_enable_action_type(button, RATBAG_BUTTON_ACTION_TYPE_KEY);
+}
+
+void
+asus_setup_resolution(struct ratbag_device *device, struct ratbag_resolution *resolution)
+{
+	struct dpi_range *dpirange;
+	dpirange = ratbag_device_data_asus_get_dpi_range(device->data);
+	if (!dpirange)
+		return;
+
+	ratbag_resolution_set_dpi_list_from_range(
+		resolution, dpirange->min, dpirange->max);
+}
+
+void
+asus_setup_led(struct ratbag_device *device, struct ratbag_led *led)
+{
+	led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+	led->type = ASUS_LED_TYPE[led->index];
+	ratbag_led_set_mode_capability(led, RATBAG_LED_ON);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_CYCLE);
+	ratbag_led_set_mode_capability(led, RATBAG_LED_BREATHING);
+}
+
+int
+asus_save_profile(struct ratbag_device *device)
+{
+	int r;
+	union asus_response response = {0};
+	union asus_request request = {
+		.data.cmd = ASUS_CMD_SAVE,
+	};
+
+	r = asus_query(device, &request, &response);
+	if (r)
+		return r;
+
+	return 0;
+}
+
+int
+asus_get_profile_data(struct ratbag_device *device, struct asus_profile_data *data)
+{
+	int r;
+	uint32_t quirks = ratbag_device_data_asus_get_quirks(device->data);
+	union asus_response response = {0};
+	union asus_request request = {
+		.data.cmd = ASUS_CMD_GET_PROFILE_DATA,
+	};
+
+	r = asus_query(device, &request, &response);
+	if (r)
+		return r;
+
+	if (quirks & ASUS_QUIRK_STRIX_PROFILE)
+		data->profile_id = response.data.results[7];
+	else
+		data->profile_id = response.data.results[8];
+
+	// primary: major, minor, build
+	data->version_primary[0] = response.data.results[13];
+	data->version_primary[1] = response.data.results[12];
+	data->version_primary[2] = response.data.results[11];
+
+	// secondary: major, minor, build
+	data->version_secondary[0] = response.data.results[4];
+	data->version_secondary[1] = response.data.results[3];
+	data->version_secondary[2] = response.data.results[2];
+
+	return 0;
+}
+
+int
+asus_set_profile(struct ratbag_device *device, unsigned int index)
+{
+	int r;
+	union asus_response response = {0};
+	union asus_request request = {
+		.data.cmd = ASUS_CMD_SET_PROFILE,
+		.data.params[0] = index,
+	};
+
+	r = asus_query(device, &request, &response);
+	if (r)
+		return r;
+
+	return 0;
+}
+
+// read button bindings using ButtonMapping configuration property
+int
+asus_get_button_data(struct ratbag_device *device, union asus_button_data *data)
+{
+	int r;
+	union asus_response response = {0};
+	union asus_request request = {
+		.data.cmd = ASUS_CMD_GET_BUTTON_DATA,
+	};
+
+	r = asus_query(device, &request, &response);
+	if (r)
+		return r;
+
+	memcpy(data->raw, response.raw, sizeof(data->raw));
+	return 0;
+}
+
+// set button binding using ASUS code of the button
+int
+asus_set_button_action(struct ratbag_device *device, unsigned int asus_index, uint8_t asus_code, uint8_t asus_type)
+{
+	int r;
+	union asus_response response = {0};
```

Comment:
unless I overlooked something, you could `memset` the response in `asus_query` which would reduce the chance of a bug being introduced by forgetting the `= {0};`.